jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/server/ServerCloseTest.java

```diff
@@ -0,0 +1,278 @@
+    }
+
+    @AfterEach
+    public void stopServer() throws Exception
+    {
+        server.stop();
+    }
+
+    @BeforeEach
+    public void startClient() throws Exception
+    {
+        client = new WebSocketClient();
+        client.setIdleTimeout(Duration.ofSeconds(2));
+        client.start();
+    }
+
+    @AfterEach
+    public void stopClient() throws Exception
+    {
+        client.stop();
+    }
+
+    private void close(Session session)
+    {
+        if (session != null)
+        {
+            session.close();
+        }
+    }
+
+    /**
+     * Test fast close (bug #403817)
+     *
+     * @throws Exception on test failure
+     */
+    @Test
+    public void fastClose() throws Exception
+    {
+        ClientUpgradeRequest request = new ClientUpgradeRequest();
+        request.setSubProtocols("fastclose");
+        CloseTrackingEndpoint clientEndpoint = new CloseTrackingEndpoint();
+
+        URI wsUri = WSURI.toWebsocket(server.getURI().resolve("/ws"));
+        Future<Session> futSession = client.connect(clientEndpoint, wsUri, request);
+
+        Session session = null;
+        try
+        {
+            session = futSession.get(5, SECONDS);
+
+            // Verify that client got close
+            clientEndpoint.assertReceivedCloseEvent(5000, is(StatusCode.NORMAL), containsString(""));
+
+            // Verify that server socket got close event
+            AbstractCloseEndpoint serverEndpoint = serverEndpointCreator.pollLastCreated();
+            assertThat("Fast Close Latch", serverEndpoint.closeLatch.await(5, SECONDS), is(true));
+            assertThat("Fast Close.statusCode", serverEndpoint.closeStatusCode, is(StatusCode.ABNORMAL));
+        }
+        finally
+        {
+            close(session);
+        }
+    }
+
+    /**
+     * Test fast fail (bug #410537)
+     *
+     * @throws Exception on test failure
+     */
+    @Test
+    public void fastFail() throws Exception
+    {
+        ClientUpgradeRequest request = new ClientUpgradeRequest();
+        request.setSubProtocols("fastfail");
+        CloseTrackingEndpoint clientEndpoint = new CloseTrackingEndpoint();
+
+        URI wsUri = WSURI.toWebsocket(server.getURI().resolve("/ws"));
+        Future<Session> futSession = client.connect(clientEndpoint, wsUri, request);
+
+        Session session = null;
+        try(StacklessLogging ignore = new StacklessLogging(FastFailEndpoint.class, WebSocketSessionImpl.class))
+        {
+            session = futSession.get(5, SECONDS);
+
+            // Verify that client got close indicating SERVER_ERROR
+            clientEndpoint.assertReceivedCloseEvent(5000, is(StatusCode.SERVER_ERROR), containsString("Intentional FastFail"));
+
+            // Verify that server socket got close event
+            AbstractCloseEndpoint serverEndpoint = serverEndpointCreator.pollLastCreated();
+            serverEndpoint.assertReceivedCloseEvent(5000, is(StatusCode.SERVER_ERROR), containsString("Intentional FastFail"));
```

Comment:
I changed line 180 to use `containsString("")` to reach this line.

This is failing due to a timeout on the `closeLatch`, the problem is that in `JettyWebSocketFrameHandler` the `closeHandle` is only invoked when a close frame is received, this is incorrect and should be instead invoked in the `onClosed(CloseStatus, Callback)` method

Reply:
Same as above.

Reply:
this failed for a different reason than above so its not the same, this is due to a problem with the implementation of the `JettyWebSocketFrameHandler`, and its probably the reason why a lot of the tests are failing